CRM/Contact/BAO/Contact/Permission.php

```diff
@@ -99,7 +202,8 @@ public static function cache($userID, $type = CRM_Core_Permission::VIEW, $force
     }
 
     if (!$force) {
-      if (!empty($_processed[$userID])) {
+      // skip if already calculated
+      if (!empty($_processed[$type][$userID])) {
```

Comment:
It might be better to do a select from the table to check there are entries for the user id - I'm worried the cache could be truncated from out of under someone